core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java

```diff
@@ -225,13 +240,9 @@ public synchronized void process(Iterable<QueryExecution> runningQueries, Suppli
         if (!(lowMemoryKiller instanceof NoneLowMemoryKiller) &&
                 outOfMemory &&
                 !queryKilled &&
-                nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) > 0) {
-            if (isLastKillTargetGone(runningQueries)) {
-                callOomKiller(runningQueries);
-            }
-            else {
-                log.debug("Last killed target is still not gone: %s", lastKillTarget);
-            }
+                nanosSince(lastTimeNotOutOfMemory).compareTo(killOnOutOfMemoryDelay) > 0 &&
+                lastKillTarget.isEmpty() && nanosSince(lastTimeKillCompleted).compareTo(killOnOutOfMemoryAfterKillDelay) > 0) {
```

Comment:
Why is the extra delay needed? Doesn't `isLastKillTargetGone` protect the OOM killer from being invoked prematurely?

Reply:
It is problematic when we are working with a query with task level fault tolerance.

The list of tasks and pool memory info are harvested on the worker side independently (those are held by two different structures).
And you can get the memory counter saying "pool is full" and high memory usage for query Q. While the list of tasks for query Q on the node empty (as tasks were just killed and are already marked as failed).

In this scenario previously we ended up killing whole query, even though query was running with task-based retries (the information that the query uses task-based retries was derived from existence of list of tasks for query in `MemoryInfo.tasksMemoryInfo`).
With last commit from this PR ([Do not kill whole queries with task retries enabled on oom](https://github.com/trinodb/trino/pull/11800/commits/d60e113dccb8b2b72fe39f2285ac9a26f4fe9595)) we would no longer whole fault-tolerant query. But without extra delay, we would still assume the node is blocked on memory, and kill something else on this node. Hence extra delay is useful here.

Reply:
I'm trying to think about a scenario when nodes are not getting blocked on memory all at once. I'm afraid that the extra delay may slow down scheduling for large clusters (100 - 1000 nodes).

From what i understand the last commit from this PR should prevent an entire fault tolerant query from being killed. However there's still a chance that a task might get killed unnecessarily if the memory pool reports that it is still blocked but the tasks are already finished. I wonder how difficult is it to return a consistent information, when the memory pool reservation is consistent with a list of tasks?

Reply:
> I'm trying to think about a scenario when nodes are not getting blocked on memory all at once. I'm afraid that the extra delay may slow down scheduling for large clusters (100 - 1000 nodes).

If we consider cluster oom a common thing - delay may slow done things surely. The question is should we consider it a common thing. Task killing itself also slows things down so we should rather try to minimize that.

> 
> From what i understand the last commit from this PR should prevent an entire fault tolerant query from being killed. However there's still a chance that a task might get killed unnecessarily if the memory pool reports that it is still blocked but the tasks are already finished. I wonder how difficult is it to return a consistent information, when the memory pool reservation is consistent with a list of tasks?

It is exactly as you write. 
It does not look it is possible to get these two pieces of information consistently. They come from very different places with no shared synchronization paradigms:
* memory info we get from `MemoryPoolInfo.getInfo` (via `LocalMemoryManager`)
* task info we get from `SqlTaskManager`

Reply:
Should be good now.
PTAL @arhimondr